### ISSUE TITLE
Support for Parallel Processing of files

### ISSUE THREAD

**Issue by rudrakshkarpe** [microsoft/markitdown#135]
## Description
Currently, PDF generation with markitdown processes pages sequentially, leading to longer rendering times. By adding the ability to process PDF pages in parallel, we can significantly improve performance and reduce overall generation time. Please consider implementing parallel page processing support.

## PDF Conversion Performance

| File                  | Size  | Pages | Time (seconds) |
|-----------------------|-------|-------|----------------|
| Test_parsing_50MB.pdf | 50MB  | 1113  | 34.64          |
| Test_parsing_20MB.pdf | 20MB  | 598 | 9.23       |
| Test_parsing_20MB.pdf | 2MB  | 23 | 0.27       |

**Comment by AliHaider20** [comment]
Were you using LLM client?

**Comment by rudrakshkarpe** [comment]
Nope, I used it with the without LLM client. I agree on the deadlock issue still is there way we can process the pages parallelly or support by library so, that we can markdown can be considered over the PyMuPDF or other similar fast processing libraries?

**Comment by AliHaider20** [comment]
I see that PyMuPDF is faster, but lacks in performance. But, yeah I just looked at their code and there is a possibility of parallel processing multiple documents and not pages. For processing multiple pdf_miner_6 should support it.

**Comment by sqrt676** [comment]
I agree with @rudrakshkarpe on this. This will help in digitization pipeline for heavy doc load. We need to add a flag and add support for this as a service layer.

**Comment by prateekralhan** [comment]
@rudrakshkarpe , @sqrt676 why don't you use pyspark for doing this if you are trying to do this on a big document volume corpus?
Just wrap your python function for MD conversion as a pyspark UDF and use it.

Here is a very basic example: https://www.geeksforgeeks.org/convert-python-functions-into-pyspark-udf/

**Comment by rudrakshkarpe** [comment]
> [@rudrakshkarpe](https://github.com/rudrakshkarpe) , [@sqrt676](https://github.com/sqrt676) why don't you use pyspark for doing this if you are trying to do this on a big document volume corpus? Just wrap your python function for MD conversion as a pyspark UDF and use it.
> 
> Here is a very basic example: https://www.geeksforgeeks.org/convert-python-functions-into-pyspark-udf/

@prateekralhan I am afraid the pyspark operation will add more latency in the process?

**Comment by prateekralhan** [comment]
@rudrakshkarpe , In terms of what python offers natively this would be still far more performant.

If you still feel that you need more optimization, you can use pandas_udf() based functionality and you need to ensure your spark executors have enough memory to perform these operations based on document corpus' size.

**Comment by rudrakshkarpe** [comment]
Thanks @prateekralhan, I'll try with this approach

**Comment by arubenruben** [comment]
> Were you using the LLM client? Even if not that is not much of a time. As per my knowledge the parallel processing is not a possibility in Python due to Deadlock issue.

@AliHaider20  Do you have any references/sources to support this claim?
I suspect my FastAPI wrapper around Markitdown is facing a deadlock when I activate multiple workers. The code does not have enough complexity for me to suffer from such problems, which indicates that might be something on the library side.

If you have any reference to this markitdown limitation, I would be very glad. Thanks.

**Comment by AliHaider20** [comment]
@arubenruben https://medium.com/@jaklimoff/understanding-deadlocks-in-python-with-examples-20b5e00f1eb8
